### PR TITLE
gulp release

### PR DESCRIPTION
**Pull request by jpoon**
**What this PR does / why we need it**:

Migrates to gulp 4.0 and introduces `gulp release` command which will: 

1. bump the package version
2. create the changelog. the categorization of the issues is picked up from the GH labels (kind/bug, kind/feature, etc); which is why you may have been seeing me label a bunch of issues recently.
3. create the git commit with the above changes
4. create a git tag

Example usage:

```
gulp release --semver patch --githubToken [TOKEN]
```

Once run, to push a release to remote: `git push --follow-tags`.

In the migration to gulp4.0, simplified some of the gulp tasks, we now have the following:

```
[12:09:42] Tasks for /Volumes/Source/Vim/gulpfile.js
[12:09:42] ├── tsc
[12:09:42] ├── tslint
[12:09:42] ├── prettier
[12:09:42] ├── forceprettier
[12:09:42] ├── test
[12:09:42] ├─┬ build
[12:09:42] │ └─┬ <series>
[12:09:42] │   ├── prettier
[12:09:42] │   └─┬ <parallel>
[12:09:42] │     ├── tsc
[12:09:42] │     └── tslint
[12:09:42] ├─┬ release
[12:09:42] │ └─┬ <series>
[12:09:42] │   ├── updateVersion
[12:09:42] │   ├── createChangelog
[12:09:42] │   ├── createGitCommit
[12:09:42] │   └── createGitTag
[12:09:42] └─┬ default
[12:09:42]   └─┬ <series>
[12:09:42]     ├─┬ build
[12:09:42]     │ └─┬ <series>
[12:09:42]     │   ├── prettier
[12:09:42]     │   └─┬ <parallel>
[12:09:42]     │     ├── tsc
[12:09:42]     │     └── tslint
[12:09:42]     └── test
```

Introduced a `gulp build` command which will run prettier, tsc, tslint. `gulp` (default) now runs *both* build and test. 

**Which issue(s) this PR fixes**

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by xconverge** [comment]
Is there a way to add a custom part to the changelog with this workflow?

**Comment by jpoon** [comment]
No, it's all automated and driven from the GitHub issues.

File: gulpfile.js [hunk 1]
```diff
@@ -1,25 +1,142 @@
...
+  });
+}
+
+function createChangelog(done) {
+  const imageName = 'jpoon/github-changelog-generator';
+  const version = require('./package.json').version;
+
+  var options = minimist(process.argv.slice(2), releaseOptions);
+
+  if (!options.githubToken) {
```
Comment (xconverge) [inline gulpfile.js]: any chance you can make this take an environment variable too if the argument is not defined? It is not a big deal but I had been using CHANGELOG_GITHUB_TOKEN

**Comment by xconverge** [comment]
Looks good to me, my only concern is the lack of an ability to add even a single sentence to a changelog. Breaking changes and big changes then cannot be mentioned clearly.

**Comment by jpoon** [comment]
> Breaking changes and big changes then cannot be mentioned clearly.

All driven through labels. So if you add a "backwards-incompatible" label, it'll put it in the "breaking changes" section. Similar with enhancements/features vs bugs.

File: gulpfile.js (hunk 1 above)
Comment (jpoon) [reply gulpfile.js]: Done

**Pull request merged**